Pb/catch stacktrace

peterboncz opened the pull request:
enable printing a stacktrace on crash in testing (when catch catches a fatal signal)

This was developed at MotherDuck as we noted irregular crashes (with a signal) in unittest in CI, that were not reproducing outside CI. Hence we desired a way to run unittest in CI that would provide at least a stack trace when catch.hpp would handle a crash. This change eventually allowed to fix a longer standing bug in MotherDuck.
  
changes made in this DuckDB PR:

- created a FormatStackTrace() method in common/exception.hpp
    
- modified catch.hpp to call a CATCH_STACKTRACE() macro on the reportFatal message, that adds a stacktrace to that message. Switches off sigaltstack() in that case (sigaltstack is not really needed and on MacOS, backtrace() just does not work when it has been used). If DUCKDB_DEBUG_STACKTRACE is enabled we include `duckdb/common/exception.hpp` and defines CATCH_STACKTRACE to call FormatStackTrace() 

- added the DEBUG_STACKTRACE variable in CMakeLists.txt and the Makefile 

To generate a crash please do an edit like the below example in: test/sql/storage/test_checksum.cpp

```
 30         int *crash = (int*) -1; fs->Write(*handle, &value, sizeof(int8_t), crash[0]*100);
```
and build & run as follows:
```
DEBUG_STACKTRACE=1 make release    
./build/release/test/unittest "Test functioning of checksum"
```
this will then crash with a stack trace.    

I am open to suggestions how to test this in CI. Currently it will not trigger in DuckDB CI, because DEBUG_STACKTRACE is not enabled by default there. Note that on Windows DuckDB does not support DUCKDB_DEBUG_STACKTRACE (i.e. stack-trace printing in src/common/exception.cpp), which would be a feature that would have to be developed before doing so (or would have to be  circumvented with some cmake testing).

I make no effort at function name demangling for the moment. When you are debugging rare crashes that only happen in CI, this is not the main concern.

Another limitation I have not been able to overcome is that compiling with address sanitizer (on MacOS) inhibits the stackframe printing from a signal. HenceI integrated the option now only in the release build, as that runs without address sanitizer.

CI for MotherDuck runs release/unittest (and it runs on x86+Linux anyway, and no Windows), so to MotherDuck having this PR merged would be quite suitable as MotherDuck would then just always run its CI with stacktrace printing.

peterboncz committed:
enable printing a stacktrace on crash, in testing (when catching a signal)

- created a FormatStackTrace() method in common/exception.hpp

- modified catch.hpp to call a CATCH_STACKTRACE() macro on the reporatFatal
  message, that adds a stacktrace to that method.
  Switches off sigaltstack in that case (sigaltstack is not relly needed and
  on MacOS, getting stacktraces just does not work when it has been used).

- introduced catch_stacktrace.hpp
  it defines CATCH_STACKTRACE to call FormatStackTrace()
  then includes catch.hpp

- all the other changes are just: #include "catch{,_stacktrace}.hpp"
  except test/sql/storage/test_checksum.cpp
  which I make to crash to demonstrate the effect. This change is *not*
  intended to be merged. It is to observe whether a stacktrace is printed.
  I am open to suggestions how to test this.

  ./build/release/test/unittest "Test functioning of checksum"

A limitation I have not been able to overcome is that compiling with address
sanitizer (on MacOS) also inhibits the stackframe printing. Hence I enable
it now on the release build (CI for MotherDuck runs release/unittest so
to us this would be suitable).

peterboncz committed:
Revert "enable printing a stacktrace on crash, in testing (when catching a signal)"

This reverts commit fcf571f4c2cea707a195d07c484d23b66e12a53b.

peterboncz committed:
enable printing a stacktrace on crash in testing
(when catch catches a fatal signal)

- created a FormatStackTrace() method in common/exception.hpp

- modified catch.hpp to call a CATCH_STACKTRACE() macro on the reportFatal
  message, that adds a stacktrace to that message.
  Switches off sigaltstack() in that case (sigaltstack is not really needed
  and on MacOS, backtrace() just does not work when it has been used).

- introduced catch_stacktrace.hpp
  it defines CATCH_STACKTRACE to call FormatStackTrace()
  then includes catch.hpp

- all the other changes are just: #include "catch{,_stacktrace}.hpp"

  except: test/sql/storage/test_checksum.cpp

which I make to crash to demonstrate the effect. This last change is *not*
intended to be merged. It is to observe whether a stacktrace is printed.
I am open to suggestions how to test this. Now you can see it with:

  ./build/release/test/unittest "Test functioning of checksum"

A limitation I have not been able to overcome is that compiling with address
sanitizer (on MacOS) also inhibits the stackframe printing. Hence I enable
it now on the release build (CI for MotherDuck runs release/unittest so
to us this would be suitable).

I make no effort at function name demangling for the moment. When you are
debugging rare crashes that only happen in CI, this is not the main concern.

peterboncz committed:
add missing file

peterboncz committed:
revert on-purpose crash in test/sql/storage/test_checksum.cpp

that shows a stacktrace from a SEGV when doing:

./build/release/test/unittest "Test functioning of checksum"

please revert to the previous commit in this PR to see that

peterboncz committed:
git fomat-fix, sigh

peterboncz committed:
dummy changeto trigger CI

peterboncz committed:
- removed the newlinethat format-check thinks is objectionable
- discovered that my git merge in odbc was incorrect (but why did I not spot it failed to build?)

peterboncz committed:
Merge remote-tracking branch 'origin/master' into pb/catch_stacktrace

peterboncz committed:
remove DEBUG_STACKTRACE from standard release compile flags since it turns out that this is a DuckDB freature unavailable on Windows.

peterboncz committed:
- only enable the stacktrace macro for catch.hpp if DUCKDB_DEBUG_STACKTRACE is enabled
- added DEBUG_STACKTRACE flag to Makefile, so a shell variable gets passed to cmake

peterboncz committed:
initialize new flag

Mytherin reviewed (commented):
Thanks for the PR! Looks good. Some comments below. Perhaps we also want to enable this on some CI runs, both to ensure this works and to get stacktrace information there when something happens?

peterboncz commented on src/include/duckdb/common/catch_stacktrace.hpp:
```diff
@@ -0,0 +1,9 @@
+#define CATCH_STACKTRACE(X) duckdb::Exception::FormatStackTrace(X).c_str()
+#endif
+
+#include "catch.hpp"
```
good idea, will do

peterboncz replied:
MotherDuck will enable this on its CI, which as I mentioned is a DuckDB release/unittest

notes:

- DUCKDB_DEBUG_STACKTRACE is not implemented for Windows (see src/common/exception.cpp) so do not try to build it on that platform as it will fail compilation

- this will work on release builds for Linux and MacOS (and maybe also in debug mode on x86/gcc/Linux, as I mentioned on MacOS address sanitising causes the backtrace to no longer work). But the chance of getting a crash is I think generally higher with optimised code. So that would also be an argument to prepare a release CI run.

peterboncz committed:
implemented som changes after review:

- added the catch.hpp changes inside the file itself (so all include changes are no longer required)
- moves DEBUG_STACKTRACE in Makefile to EXTENSIONS so it is passed to all targets

Mytherin reviewed (commented):
Thanks for the changes! The code looks good to me - can we perhaps also enable this in the `LinuxRelease` workflow by adding the environment variable [here](https://github.com/duckdb/duckdb/blob/master/.github/workflows/LinuxRelease.yml#L37) to make sure this is tested/used as well?

peterboncz committed:
enable DEBUG_STACKTRACE in the LinuxRelease workflow

peterboncz commented:
thanks & I pushed that change to the yml

peterboncz committed:
just changed the order (the previous CI job got canceled??)

peterboncz committed:
Merge branch 'master' into pb/catch_stacktrace

peterboncz commented:
this thing works fine in CI now, I think it's done

Mytherin commented:
Thanks! LGTM
